### ISSUE TITLE
Tests sometimes fail due to sites failing to start with "port already in use"

### ISSUE THREAD

**Issue by DamianEdwards** [dotnet/aspire#4037]
I've seen a few CI test failures this past week due to resources failing to start during tests due to the randomly assigned port being in use. Maybe we should think about ways to reduce the likelihood of this? @ReubenBond 

Example:
```
2024-04-30T17:06:05.3505825+00:00 Resource 'frontend' log: 2024-04-30T17:06:05.0981488 \x1b[41m\x1b[30mfail\x1b[39m\x1b[22m\x1b[49m: Microsoft.Extensions.Hosting.Internal.Host[11]
 [202](https://github.com/dotnet/aspire-samples/actions/runs/8898768635/job/24436629702?pr=247#step:10:203)4-04-30T17:06:05.3505879+00:00 Resource 'frontend' log:       Hosting failed to start
 2024-04-30T17:06:05.3505907+00:00 Resource 'frontend' log:       System.IO.IOException: Failed to bind to address [http://127.0.0.1:32791:](http://127.0.0.1:32791/) address already in use.
 2024-04-30T17:06:05.3505945+00:00 Resource 'frontend' log:        ---> Microsoft.AspNetCore.Connections.AddressInUseException: Address already in use
 2024-04-30T17:06:05.3505981+00:00 Resource 'frontend' log:        ---> System.Net.Sockets.SocketException (98): Address already in use
 2024-04-30T17:06:05.3506024+00:00 Resource 'frontend' log:          at System.Net.Sockets.Socket.DoBind(EndPoint endPointSnapshot, SocketAddress socketAddress)
 2024-04-30T17:06:05.3506063+00:00 Resource 'frontend' log:          at System.Net.Sockets.Socket.Bind(EndPoint localEP)
 2024-04-30T17:06:05.3506095+00:00 Resource 'frontend' log:          at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.SocketTransportOptions.CreateDefaultBoundListenSocket(EndPoint endpoint)
 2024-04-30T17:06:05.3506146+00:00 Resource 'frontend' log:          at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.SocketConnectionListener.Bind()
 2024-04-30T17:06:05.3506184+00:00 Resource 'frontend' log:          --- End of inner exception stack trace ---
 2024-04-30T17:06:05.3506214+00:00 Resource 'frontend' log:          at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.SocketConnectionListener.Bind()
 2024-04-30T17:06:05.3506251+00:00 Resource 'frontend' log:          at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.SocketTransportFactory.BindAsync(EndPoint endpoint, CancellationToken cancellationToken)
 2024-04-30T17:06:05.3506528+00:00 Resource 'frontend' log:          at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure.TransportManager.BindAsync(EndPoint endPoint, ConnectionDelegate connectionDelegate, EndpointConfig endpointConfig, CancellationToken cancellationToken)
 2024-04-30T17:06:05.3506599+00:00 Resource 'frontend' log:          at Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerImpl.<>c__DisplayClass28_0`1.<<StartAsync>g__OnBind|0>d.MoveNext()
 2024-04-30T17:06:05.3506658+00:00 Resource 'frontend' log:       --- End of stack trace from previous location ---
 2024-04-30T17:06:05.3506688+00:00 Resource 'frontend' log:          at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.BindEndpointAsync(ListenOptions endpoint, AddressBindContext context, CancellationToken cancellationToken)
 2024-04-30T17:06:05.3506737+00:00 Resource 'frontend' log:          --- End of inner exception stack trace ---
 2024-04-30T17:06:05.3506767+00:00 Resource 'frontend' log:          at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.BindEndpointAsync(ListenOptions endpoint, AddressBindContext context, CancellationToken cancellationToken)
 2024-04-30T17:06:05.3506824+00:00 Resource 'frontend' log:          at Microsoft.AspNetCore.Server.Kestrel.Core.LocalhostListenOptions.BindAsync(AddressBindContext context, CancellationToken cancellationToken)
 2024-04-30T17:06:05.3506869+00:00 Resource 'frontend' log:          at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.AddressesStrategy.BindAsync(AddressBindContext context, CancellationToken cancellationToken)
 2024-04-30T17:06:05.3506924+00:00 Resource 'frontend' log:          at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.BindAsync(ListenOptions[] listenOptions, AddressBindContext context, Func`2 useHttps, CancellationToken cancellationToken)
 2024-04-30T17:06:05.3506978+00:00 Resource 'frontend' log:          at Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerImpl.BindAsync(CancellationToken cancellationToken)
 2024-04-30T17:06:05.3507018+00:00 Resource 'frontend' log:          at Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerImpl.StartAsync[TContext](IHttpApplication`1 application, CancellationToken cancellationToken)
 2024-04-30T17:06:05.3507074+00:00 Resource 'frontend' log:          at Microsoft.AspNetCore.Hosting.GenericWebHostService.StartAsync(CancellationToken cancellationToken)
 2024-04-30T17:06:05.3507115+00:00 Resource 'frontend' log:          at Microsoft.Extensions.Hosting.Internal.Host.<StartAsync>b__15_1(IHostedService service, CancellationToken token)
 2024-04-30T17:06:05.3507154+00:00 Resource 'frontend' log:          at Microsoft.Extensions.Hosting.Internal.Host.ForeachService[T](IEnumerable`1 services, CancellationToken token, Boolean concurrent, Boolean abortOnFirstException, List`1 exceptions, Func`3 operation)
```

**Comment by DamianEdwards** [comment]
/Cc @karolz-ms for thoughts on how we might make the random port assignment DCP does more reliable.

**Comment by karolz-ms** [comment]
@DamianEdwards are the tests in question launching multiple app host instances in parallel?

**Comment by DamianEdwards** [comment]
@karolz-ms no. There are only two theories and they're in the same class using xUnit default parallelism settings, so should all be sequential, one at a time.

**Comment by karolz-ms** [comment]
OK. I guess it is time for me to take a closer look.

**Comment by DamianEdwards** [comment]
Got another hit:
`2024-05-02T17:56:24.1560247+00:00 Resource 'angular' log: ? Port 36791 is already in use.`

**Comment by karolz-ms** [comment]
I did spend some time looking at how the port allocation works. We made a change that should eliminate port conflicts if multiple Aspire apps/tests are running https://github.com/microsoft/usvc-apiserver/pull/369 

Ultimately though DCP is doing the port check, but a service is doing socket creation, so there is always TOCTOU gap that we cannot fully close. There is always a chance that some other program on the system will grab the port that DCP selected to be used by the app.
